### ISSUE TITLE
The dropout layer is not disabled after setting the Module to Eval.

### ISSUE THREAD

**Issue by Leo-LiHao** [dotnet/TorchSharp#549]
Here is the code for reproducing the bug.

### Approach 1

Using a `Sequential` to store `Modules`

```C#
class Reproduce : Module
{
    private readonly Module layers;
    public Reproduce(string name) : base(name)
    {
        var modules = new List<(string, Module)>();
        modules.Add(("fc_linear_1", Linear(100, 7)));
        modules.Add(("fc_dropout_1", Dropout(0.5)));
        layers = Sequential(modules);
    }
    
    // Following code can work for version 0.96.3, not work for version 0.93.1 
    //public override void eval()
    //{
    //    base.eval();
    //    layers.eval();
    //}

    public override Tensor forward(Tensor input)
    {
        return layers.forward(input);
    }
}
```

### Approach 2

```C#
class Reproduce2 : Module
{
    private readonly Module layers;
    private Module dropout;
    private Module dense;
    public Reproduce2(string name) : base(name)
    {
        dense = Linear(100, 7);
        dropout = Dropout(0.5);
    }

    public override Tensor forward(Tensor input)
    {
        return dropout.forward(dense.forward(input));
    }
}
```

### Results for Approach 1

```C#
torch.random.manual_seed(0);
var x = torch.arange(0, 100, 1, torch.float32, torch.CPU).reshape(1, 100);

var model = new Reproduce("reproduce");
model.eval();
Console.WriteLine("Reproducing 1...");
Tensor y;
for (var run = 0; run < 3; run++)
{
    Console.WriteLine($"{run}th running...");
    y = model.forward(x);
    for (var i = 0; i < 7; i++)
    {
        Console.Write($"{y.ReadCpuSingle(i)} ");
    }
    Console.WriteLine("");
}
```

Version `0.96.3` with override `eval` method:
```
Reproducing 1...
0th running...
-1.7911556 -39.207882 -24.935097 61.1867 -8.499707 14.684342 26.447653 
1th running...
-1.7911556 -39.207882 -24.935097 61.1867 -8.499707 14.684342 26.447653 
2th running...
-1.7911556 -39.207882 -24.935097 61.1867 -8.499707 14.684342 26.447653 
```

Version `0.96.3` without override `eval` method:
```
Reproducing 1...
0th running...
-3.5823112 -78.415764 -0 0 -16.999414 0 0 
1th running...
-3.5823112 -0 -49.870193 122.3734 -0 29.368685 0 
2th running...
-3.5823112 -0 -49.870193 122.3734 -16.999414 29.368685 52.895306 
```

Version `0.93.1` with override `eval` method:
```
Reproducing 1...
0th running...
-3.5823007 -78.415764 -0 0 -16.999403 0 0 
1th running...
-3.5823007 -0 -49.870197 122.373405 -0 29.368685 0 
2th running...
-3.5823007 -0 -49.870197 122.373405 -16.999403 29.368685 52.895294 
```

### Results for Approach 2

```C#
Console.WriteLine("Reproducing 2...");
var model2 = new Reproduce2("reproduce");
model2.eval();
for (var run = 0; run < 3; run++)
{
    Console.WriteLine($"{run}th running...");
    y = model2.forward(x);
    for (var i = 0; i < 7; i++)
    {
        Console.Write($"{y.ReadCpuSingle(i)} ");
    }
    Console.WriteLine("");
}
```

Version `0.96.3`:

```
Reproducing 2...
0th running...
4.2551045 118.96411 -0 13.26274 -97.54702 80.192764 0 
1th running...
4.2551045 118.96411 -0 0 -97.54702 0 0 
2th running...
4.2551045 118.96411 -0 0 -97.54702 80.192764 0 
```

Version `0.93.1`:

```
Reproducing 2...
0th running...
-1.0717726 93.69564 -0 2.073273 -83.66095 47.344925 0 
1th running...
-0 93.69564 -81.99559 0 -0 47.344925 99.293015 
2th running...
-0 0 -0 2.073273 -83.66095 47.344925 99.293015 
```

## Conclusion

- In `0.93.1`, the two approaches above cannot disable the `dropout` layer. 
- In `0.96.3`, Approach 2 cannot disable the `dropout` layer, Approach 1 has to manually override the `eval` method to disable the `dropout` layer. 

PS: Don't forget to change `eval` to `Eval` in version `0.93.1`

**Comment by Leo-LiHao** [comment]
> @Leo-LiHao: from what I can tell, the problem (in 0.96.3) is that your model constructor doesn't have a RegisterComponents() call after the construction of all submodules.
> 
> When I change the ctor to look like this:
> 
> ```cs
> public Reproduce(string name) : base(name)
> {
>     var modules = new List<(string, Module)>();
>     modules.Add(("fc_linear_1", Linear(100, 7)));
>     modules.Add(("fc_dropout_1", Dropout(0.5)));
>     layers = Sequential(modules);
>     RegisterComponents();
> }
> ```
> 
> It seems like it's doing the right thing on my system.

Thank you. It works in 0.96.3. However, I can still reproduce that in 0.93.1. Any idea about which PR fixes this issue?

**Comment by NiklasGustafsson** [comment]
> Thank you. It works in 0.96.3. However, I can still reproduce that in 0.93.1 (and also 0.93.9). Any idea about which PR fixes this issue?

Bug #500, fixed in PR #501
